lib/auth/password.go

```diff
@@ -178,23 +257,110 @@ func (s *AuthServer) getOTPType(user string) (string, error) {
 	return teleport.HOTP, nil
 }
 
-// GetOTPData returns the OTP Key, Key URL, and the QR code.
-func (s *AuthServer) GetOTPData(user string) (string, []byte, error) {
-	// get otp key from backend
-	otpSecret, err := s.GetTOTP(user)
+func (s *AuthServer) changePasswordWithToken(req ChangePasswordWithTokenRequest) (services.User, error) {
```

Comment:
Can group some of the logic and add comments within? Makes it easier to follow. Something like:

```go
func (s *AuthServer) changePasswordWithToken(req ChangePasswordWithTokenRequest) (services.User, error) {
	// Get cluster configuration and check if local auth is allowed.
	clusterConfig, err := s.GetClusterConfig()
	if err != nil {
		return nil, trace.Wrap(err)
	}
	if clusterConfig.GetLocalAuth() == false {
		return nil, trace.AccessDenied(noLocalAuth)
	}

	// Check if the signup token exists.
	userToken, err := s.GetUserToken(req.TokenID)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	if userToken.Expiry().Before(s.clock.Now().UTC()) {
		return nil, trace.BadParameter("expired token")
	}

	// Check the password meets the complexity requirements and update it.
	err = services.VerifyPassword(req.Password)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	err = s.UpsertPassword(username, []byte(req.Password))
	if err != nil {
		return nil, trace.Wrap(err)
	}

	[...]
}
```

Reply:
Currently it has the flow which minimizes the chances of partially updated user + valid token. I will add more comments.